### PR TITLE
Implement FRIFOLD4 operation/instruction

### PR DESCRIPTION
**Pull request by bobbinth**
This PR adds a new operation and associated instruction to the VM to perform FRI layer folding in a single VM cycle (`FRIE2F4`). The folding assumes that we are in the degree 2 extension of the base field and folding factor is 4.

- [x] Implement basic `FRIE2F4` VM operation.
- [x] Implement `fri_ext2fold4` assembly instruction.
- [x] Add integration tests.
- [x] Update docs.

Left for future PRs:
- Update AIR constraints.

### CODE REVIEW

File: docs/src/design/stack/crypto_ops.md [hunk 1]
```diff
@@ -119,3 +119,34 @@ The above constraint enforces that the specified input and output rows for both,
 
 The effect of this operation on the rest of the stack is:
 * **No change** for positions starting from $4$.
+
+## FRIE2F4
+The `FRIE2F4` operation performs FRI layer folding by a factor of 4 for FRI protocol executed in a degree 2 extension of the base field. It also performs several computations needed for checking correctness of the folding from the previous layer as well as simplifying folding of the next FRI layer.
+
+The stack for the operation is expected to be arranged as follows:
+- The first $8$ stack elements contain $4$ query points to be folded. Each point is represented by two field elements because points to be folded are in the extension field. We denote these points as $q_0 = (v_0, v_1)$, $q_1 = (v_2, v_3)$, $q_2 = (v_4, v_5)$, $q_3 = (v_6, v_7)$.
+- The next element $f\_pos$ is the query position in the folded domain. It can be computed as $pos \mod n$, where $pos$ is the position in the source domain, and $n$ is size of the folded domain.
```
Comment (itzmeanjan) [inline docs/src/design/stack/crypto_ops.md:128]: Do you mean $f_{pos}$ ? If so, surround `pos` with `{}`

File: docs/src/design/stack/crypto_ops.md [hunk 2]
```diff
@@ -119,3 +119,34 @@ The above constraint enforces that the specified input and output rows for both,
...
+- The next element $d\_seg$ is a value indicating domain segment from which the position in the original domain was folded. It can be computed as $\lfloor \frac{pos}{n} \rfloor$. Since the size of the source domain is always $4$ times bigger than the size of the folded domain, possible domain segment values can be $0$, $1$, $2$, or $3$.
```
Comment (itzmeanjan) [inline docs/src/design/stack/crypto_ops.md:129]: Similarly it's probably $d_{seg}$.

**Review by itzmeanjan — COMMENTED** [review]
This is high level review. It looks good to me. I'll do an in-depth review of FRI op once again.

**Review by Al-Kindi-0 — APPROVED** [review]
Looks awesome! Thank you @bobbinth !

File: processor/src/operations/fri_ops.rs [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+// CONSTANTS
+// ================================================================================================
+
+const TWO: Felt = Felt::new(2);
+const TWO_INV: Felt = Felt::new(9223372034707292161);
+
+const DOMAIN_OFFSET: Felt = Felt::GENERATOR;
+
+// Pre-computed powers of 1/tau, where tau is the generator of multiplicative subgroup of size 2.
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/operations/fri_ops.rs:12]: Fixed!

File: docs/src/design/stack/crypto_ops.md (hunk 1 above)
Comment (bobbinth) [reply docs/src/design/stack/crypto_ops.md:128]: I actually did want an underscore there - `f_pos` as it stands for "folded position". $f_{pos}$ doesn't quite work. I guess we could change it to $pos_{folded}$ but it is a bit more verbose.

Comment (itzmeanjan) [reply docs/src/design/stack/crypto_ops.md:128]: Oh I misunderstood. Makes sense now.

**Pull request merged**